File: backend/services/stigmer-server/pkg/domain/executioncontext/temporal/activities/delete_execution_context.go
package activities

import (
	"context"
	"errors"

	"github.com/rs/zerolog/log"
	executioncontextv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/executioncontext/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/backend/libs/go/store"
)

// DeleteExecutionContextActivityImpl deletes the ExecutionContext associated with
// a completed execution (agent or workflow).
//
// The ExecutionContext is an ephemeral resource containing the fully-merged
// environment (environment_refs values overridden by runtime_env, filtered to
// the blueprint's declared env keys), including secrets. It must be cleaned up
// when the execution finishes to prevent sensitive data from persisting
// beyond the execution lifetime.
//
// This is a local activity (runs in-process on the workflow worker) shared by
// both AgentExecution and WorkflowExecution workflows.
//
// Behavior:
//   - Idempotent: no-op if the ExecutionContext does not exist
//   - Best-effort: logs errors but never returns them -- cleanup failure must
//     not affect the workflow outcome. A TTL-based backup index provides
//     defense-in-depth for orphaned contexts.
//   - Security-aware: logs variable count, never variable names or values
type DeleteExecutionContextActivityImpl struct {
	store store.Store
}

// NewDeleteExecutionContextActivityImpl creates a new activity implementation.
func NewDeleteExecutionContextActivityImpl(store store.Store) *DeleteExecutionContextActivityImpl {
	return &DeleteExecutionContextActivityImpl{store: store}
}

// DeleteExecutionContext finds and deletes the ExecutionContext for the given
// execution ID. The execution ID may be either an AgentExecution or
// WorkflowExecution ID -- the lookup uses the spec.executionId field.
func (a *DeleteExecutionContextActivityImpl) DeleteExecutionContext(ctx context.Context, executionID string) error {
	log.Debug().
		Str("execution_id", executionID).
		Msg("Cleaning up ExecutionContext for execution")

	ec := &executioncontextv1.ExecutionContext{}

	if err := a.store.FindByField(
		ctx,
		apiresourcekind.ApiResourceKind_execution_context,
		"spec.executionId",
		executionID,
		ec,
	); err != nil {
		if errors.Is(err, store.ErrNotFound) {
			log.Debug().
				Str("execution_id", executionID).
				Msg("No ExecutionContext found for execution -- nothing to clean up")
			return nil
		}
		log.Warn().
			Err(err).
			Str("execution_id", executionID).
			Msg("Failed to query ExecutionContext -- will rely on TTL cleanup")
		return nil
	}

	contextID := ec.GetMetadata().GetId()
	dataCount := len(ec.GetSpec().GetData())

	if err := a.store.DeleteResource(
		ctx,
		apiresourcekind.ApiResourceKind_execution_context,
		contextID,
	); err != nil {
		log.Warn().
			Err(err).
			Str("execution_id", executionID).
			Str("context_id", contextID).
			Msg("Failed to delete ExecutionContext -- will rely on TTL cleanup")
		return nil
	}

	log.Info().
		Str("execution_id", executionID).
		Str("context_id", contextID).
		Int("variables", dataCount).
		Msg("Deleted ExecutionContext for execution")

	return nil
}

// DeleteExecutionContextActivityName is the activity name for registration.
const DeleteExecutionContextActivityName = "DeleteExecutionContext"

File: backend/services/stigmer-server-ts/src/domain/executioncontext/temporal/delete-execution-context.ts
/**
 * DeleteExecutionContext activity seam — ports
 * pkg/domain/executioncontext/temporal/activities/delete_execution_context.go.
 *
 * The ExecutionContext is an ephemeral resource containing the
 * fully-merged environment (environment_refs values overridden by
 * runtime_env, filtered to the blueprint's declared env keys), including
 * secrets. It must be cleaned up when the execution finishes so sensitive
 * data does not persist beyond the execution lifetime.
 *
 * A SEAM, deliberately unregistered: this module has no Temporal SDK
 * dependency — it is the plain store-direct function the agent-execution
 * and workflow-execution workers (D4 #18/#21) will register as a local
 * activity under DELETE_EXECUTION_CONTEXT_ACTIVITY_NAME, shared by both
 * workflows exactly as in Go.
 *
 * Behavior (Go parity):
 *   - Idempotent: no-op if the ExecutionContext does not exist.
 *   - Best-effort: logs failures but never throws — cleanup failure must
 *     not affect the workflow outcome. A failed delete leaves the row in
 *     place with its secret values still encrypted at rest (oss#535);
 *     nothing retries it later, so the WARN is the operator's signal.
 *   - Security-aware: logs the variable count, never names or values.
 */
import { ExecutionContextSchema } from "@stigmer/protos/ai/stigmer/agentic/executioncontext/v1/api_pb";
import type { ExecutionContext } from "@stigmer/protos/ai/stigmer/agentic/executioncontext/v1/api_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import type { Logger } from "../../../boot/logger.js";
import { ResourceNotFoundError } from "../../../store/interface.js";
import type { Store } from "../../../store/interface.js";

/** The activity name for worker registration — Go's constant, byte-pinned. */
export const DELETE_EXECUTION_CONTEXT_ACTIVITY_NAME = "DeleteExecutionContext";

/**
 * Finds and deletes the ExecutionContext for the given execution ID
 * (an AgentExecution or WorkflowExecution id — the lookup uses the
 * spec.executionId field).
 */
export async function deleteExecutionContextForExecution(
  store: Store,
  logger: Logger,
  executionId: string,
): Promise<void> {
  logger.debug("Cleaning up ExecutionContext for execution", { executionId });

  let ec: ExecutionContext;
  try {
    ec = await store.findByField(
      ApiResourceKind.execution_context,
      "spec.executionId",
      executionId,
      ExecutionContextSchema,
    );
  } catch (error) {
    if (error instanceof ResourceNotFoundError) {
      logger.debug(
        "No ExecutionContext found for execution -- nothing to clean up",
        { executionId },
      );
      return;
    }
    logger.warn(
      "Failed to query ExecutionContext -- leaving cleanup to the operator",
      {
        executionId,
        error: error instanceof Error ? error.message : String(error),
      },
    );
    return;
  }

  const contextId = ec.metadata?.id ?? "";
  const dataCount = Object.keys(ec.spec?.data ?? {}).length;

  try {
    await store.deleteResource(ApiResourceKind.execution_context, contextId);
  } catch (error) {
    logger.warn(
      "Failed to delete ExecutionContext -- leaving cleanup to the operator",
      {
        executionId,
        contextId,
        error: error instanceof Error ? error.message : String(error),
      },
    );
    return;
  }

  logger.info("Deleted ExecutionContext for execution", {
    executionId,
    contextId,
    variables: dataCount,
  });
}
